src/order/upper_lower.lean

```diff
@@ -0,0 +1,299 @@
+lemma is_upper_set_univ : is_upper_set (univ : set α) := λ _ _ _, id
+lemma is_lower_set_univ : is_lower_set (univ : set α) := λ _ _ _, id
+lemma is_upper_set.compl (hs : is_upper_set s) : is_lower_set sᶜ := λ a b h hb ha, hb $ hs h ha
+lemma is_lower_set.compl (hs : is_lower_set s) : is_upper_set sᶜ := λ a b h hb ha, hb $ hs h ha
+
+lemma is_upper_set.union (hs : is_upper_set s) (ht : is_upper_set t) : is_upper_set (s ∪ t) :=
+λ a b h, or.imp (hs h) (ht h)
+
+lemma is_lower_set.union (hs : is_lower_set s) (ht : is_lower_set t) : is_lower_set (s ∪ t) :=
+λ a b h, or.imp (hs h) (ht h)
+
+lemma is_upper_set.inter (hs : is_upper_set s) (ht : is_upper_set t) : is_upper_set (s ∩ t) :=
+λ a b h, and.imp (hs h) (ht h)
+
+lemma is_lower_set.inter (hs : is_lower_set s) (ht : is_lower_set t) : is_lower_set (s ∩ t) :=
+λ a b h, and.imp (hs h) (ht h)
+
+lemma is_upper_set_Union {f : ι → set α} (hf : ∀ i, is_upper_set (f i)) : is_upper_set (⋃ i, f i) :=
+λ a b h, Exists₂.imp $ forall_range_iff.2 $ λ i, hf i h
+
+lemma is_lower_set_Union {f : ι → set α} (hf : ∀ i, is_lower_set (f i)) : is_lower_set (⋃ i, f i) :=
+λ a b h, Exists₂.imp $ forall_range_iff.2 $ λ i, hf i h
+
+lemma is_upper_set_Union₂ {f : Π i, κ i → set α} (hf : ∀ i j, is_upper_set (f i j)) :
+  is_upper_set (⋃ i j, f i j) :=
+is_upper_set_Union $ λ i, is_upper_set_Union $ hf i
+
+lemma is_lower_set_Union₂ {f : Π i, κ i → set α} (hf : ∀ i j, is_lower_set (f i j)) :
+  is_lower_set (⋃ i j, f i j) :=
+is_lower_set_Union $ λ i, is_lower_set_Union $ hf i
+
+lemma is_upper_set_sUnion {S : set (set α)} (hf : ∀ s ∈ S, is_upper_set s) : is_upper_set (⋃₀ S) :=
+λ a b h, Exists₂.imp $ λ s hs, hf s hs h
+
+lemma is_lower_set_sUnion {S : set (set α)} (hf : ∀ s ∈ S, is_lower_set s) : is_lower_set (⋃₀ S) :=
+λ a b h, Exists₂.imp $ λ s hs, hf s hs h
+
+lemma is_upper_set_Inter {f : ι → set α} (hf : ∀ i, is_upper_set (f i)) : is_upper_set (⋂ i, f i) :=
+λ a b h, forall₂_imp $ forall_range_iff.2 $ λ i, hf i h
+
+lemma is_lower_set_Inter {f : ι → set α} (hf : ∀ i, is_lower_set (f i)) : is_lower_set (⋂ i, f i) :=
+λ a b h, forall₂_imp $ forall_range_iff.2 $ λ i, hf i h
+
+lemma is_upper_set_Inter₂ {f : Π i, κ i → set α} (hf : ∀ i j, is_upper_set (f i j)) :
+  is_upper_set (⋂ i j, f i j) :=
+is_upper_set_Inter $ λ i, is_upper_set_Inter $ hf i
+
+lemma is_lower_set_Inter₂ {f : Π i, κ i → set α} (hf : ∀ i j, is_lower_set (f i j)) :
+  is_lower_set (⋂ i j, f i j) :=
+is_lower_set_Inter $ λ i, is_lower_set_Inter $ hf i
+
+lemma is_upper_set_sInter {S : set (set α)} (hf : ∀ s ∈ S, is_upper_set s) : is_upper_set (⋂₀ S) :=
+λ a b h, forall₂_imp $ λ s hs, hf s hs h
+
+lemma is_lower_set_sInter {S : set (set α)} (hf : ∀ s ∈ S, is_lower_set s) : is_lower_set (⋂₀ S) :=
+λ a b h, forall₂_imp $ λ s hs, hf s hs h
+
+lemma is_upper_set.of_dual (hs : is_upper_set s) : is_lower_set (of_dual ⁻¹' s) := hs
+lemma is_lower_set.of_dual (hs : is_lower_set s) : is_upper_set (of_dual ⁻¹' s) := hs
```

Comment:
I think @jcommelin is asking for
```suggestion
lemma is_upper_set.of_dual (hs : is_upper_set s) : is_lower_set (of_dual ⁻¹' s) := hs
lemma is_lower_set.of_dual (hs : is_lower_set s) : is_upper_set (of_dual ⁻¹' s) := hs
lemma is_upper_set.to_dual {s : set (order_dual α)} (hs : is_upper_set s) : is_lower_set (to_dual ⁻¹' s) := hs
lemma is_lower_set.to_dual {s : set (order_dual α)} (hs : is_lower_set s) : is_upper_set (to_dual ⁻¹' s) := hs
```
although I think these names might be backwards.

Reply:
I also think these might be worth having as `iff`s

Reply:
They're backward because we're stating them with the preimage. Is stating them with the preimage the right way?

Reply:
Yes, the preimage is the definitionally better direction